components/org.wso2.carbon.identity.application.authenticator.biometric/src/main/java/org/wso2/carbon/identity/application/authenticator/biometric/BiometricAuthenticator.java

```diff
@@ -0,0 +1,161 @@
+import org.wso2.carbon.identity.application.authentication.framework.FederatedApplicationAuthenticator;
+import org.wso2.carbon.identity.application.authentication.framework.context.AuthenticationContext;
+import org.wso2.carbon.identity.application.authentication.framework.exception.AuthenticationFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.application.authenticator.biometric.dao.impl.DeviceDAOImpl;
+import org.wso2.carbon.identity.application.authenticator.biometric.notification.handler.impl.FirebasePushNotificationSenderImpl;
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import java.io.IOException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import static org.wso2.carbon.identity.application.authenticator.biometric.BiometricAuthenticatorConstants.BIOMETRIC_AUTH_CHALLENGE;
+import static org.wso2.carbon.identity.application.authenticator.biometric.BiometricAuthenticatorConstants.CONTEXT_KEY;
+import static org.wso2.carbon.identity.application.authenticator.biometric.BiometricAuthenticatorConstants.DOMAIN_NAME;
+import static org.wso2.carbon.identity.application.authenticator.biometric.BiometricAuthenticatorConstants.WAIT_PAGE;
+
+/**
+ * Biometric Authenticator class.
+ */
+public class BiometricAuthenticator extends AbstractApplicationAuthenticator
+        implements FederatedApplicationAuthenticator {
+
+    private static final long serialVersionUID = 8272421416671799253L;
+    private static final Log log = LogFactory.getLog(BiometricAuthenticator.class);
+
+
+    @Override
+    public String getFriendlyName() {
+
+        return BiometricAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+        log.info("Signed challenge from the form submitted:" + request.getParameter(BiometricAuthenticatorConstants.
+                SIGNED_CHALLENGE));
+        return request.getParameter(BiometricAuthenticatorConstants.SIGNED_CHALLENGE) != null;
+    }
+
+    @Override
+    public String getContextIdentifier(javax.servlet.http.HttpServletRequest request) {
+
+        return request.getParameter(CONTEXT_KEY);
+    }
+
+    @Override
+    public String getName() {
+
+        return BiometricAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
+                                                 AuthenticationContext context) throws AuthenticationFailedException {
+
+        AuthenticatedUser user = context.getSequenceConfig().getStepMap().
```

Comment:
Can't this be null? better to check null for AuthenticatedUser

Reply:
If its null, the authentication won't proceed from that step onwards and will prompt a login error which is handled by the sp login page itself.